pkg/cmd/server/middleware.go

```diff
@@ -210,3 +214,225 @@ func (mc *MiddlewareChain[T]) modify(modifications ...MiddlewareModification[T])
+				alreadyExecuted: soeb.executed,
+				notExecuted:     soeb.notExecuted,
+			}
+			return soeb.streamInterceptor(srv, wrappedStream, info, handler)
+		},
+	}
+}
+
+func NewUnaryMiddleware() *UnaryOrderEnforcerBuilder {
+	return &UnaryOrderEnforcerBuilder{}
+}
+
+type UnaryOrderEnforcerBuilder struct {
+	name            string
+	interceptor     grpc.UnaryServerInterceptor
+	internal        bool
+	alreadyExecuted string
+	notExecuted     string
+}
+
+func (soeb *UnaryOrderEnforcerBuilder) WithName(name string) *UnaryOrderEnforcerBuilder {
+	soeb.name = name
+	return soeb
+}
+
+func (soeb *UnaryOrderEnforcerBuilder) WithInterceptor(interceptor grpc.UnaryServerInterceptor) *UnaryOrderEnforcerBuilder {
+	soeb.interceptor = interceptor
+	return soeb
+}
+
+func (soeb *UnaryOrderEnforcerBuilder) WithInternal(internal bool) *UnaryOrderEnforcerBuilder {
+	soeb.internal = internal
+	return soeb
+}
+
+func (soeb *UnaryOrderEnforcerBuilder) EnsureAlreadyExecuted(name string) *UnaryOrderEnforcerBuilder {
+	soeb.alreadyExecuted = name
+	return soeb
+}
+
+func (soeb *UnaryOrderEnforcerBuilder) EnsureNotExecuted(name string) *UnaryOrderEnforcerBuilder {
+	soeb.notExecuted = name
+	return soeb
+}
+
+func (soeb *UnaryOrderEnforcerBuilder) Done() ReferenceableMiddleware[grpc.UnaryServerInterceptor] {
+	if !inTest() {
+		return ReferenceableMiddleware[grpc.UnaryServerInterceptor]{
+			Name:       soeb.name,
+			Internal:   soeb.internal,
+			Middleware: soeb.interceptor,
+		}
+	}
+
+	return ReferenceableMiddleware[grpc.UnaryServerInterceptor]{
+		Name:     soeb.name,
+		Internal: soeb.internal,
+		Middleware: func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+			if ctx.Value(interceptorsExecuted) == nil {
+				handle := executedHandle{executed: make(map[string]struct{}, 0)}
+				ctx = context.WithValue(ctx, interceptorsExecuted, &handle)
+			}
+
+			if err := mustHaveExecuted(ctx, soeb.alreadyExecuted); err != nil {
+				return nil, err
+			}
+
+			if err := mustHaveNotExecuted(ctx, soeb.notExecuted); err != nil {
+				return nil, err
+			}
+
+			mustMarkAsExecuted(ctx, soeb.name)
+			return soeb.interceptor(ctx, req, info, handler)
+		},
+	}
+}
+
+func mustHaveNotExecuted(ctx context.Context, notExecuted string) error {
+	if notExecuted == "" {
+		return nil
+	}
+
+	val := ctx.Value(interceptorsExecuted)
+	if val == nil {
+		return fmt.Errorf("interception order validation bookkeeping not present in context")
+	}
+
+	handle := val.(*executedHandle)
+	if _, ok := handle.executed[notExecuted]; ok {
+		return fmt.Errorf("expected interceptor %s to be not already executed", notExecuted)
+	}
+
+	return nil
+}
+
+func mustHaveExecuted(ctx context.Context, expectedExecuted string) error {
+	if expectedExecuted == "" {
+		return nil
+	}
+
+	val := ctx.Value(interceptorsExecuted)
+	if val == nil {
+		panic("interception order validation bookkeeping not present in context")
```

Comment:
use spiceerrors.MustBugf

Reply:
I can add it but it does not provide value because none of this is executed in runtime

Reply:
ok